FE-1140 - [Dashboard UX Redesign] Set up page plumbing

nicklemmon opened the pull request:
[FE-1140](https://jira.int.messagesystems.com/browse/FE-1140)

### What Changed
- Adds new V2 Dashboard page with a container for handling Redux
- Adds supporting tests

### How To Test
- Enable Hibana and the account UI option `allow_dashboard_v2`
- Visit `/dashboardV2` and verify the page renders
- Disable Hibana and visit `/dashboardV2` and verify the 404 page renders
- Re-enable Hibana and remove the `allow_dashboard_v2` - visit `/dashboardV2` and verify the 404 page renders

### To Do
- [ ] Incorporate feedback

nicklemmon commented on src/pages/dashboard/DashboardPageV2.js:
```diff
@@ -0,0 +1,95 @@
+import { Box, Layout, Page, Panel, ScreenReaderOnly, Stack, Text } from 'src/components/matchbox';
+import { Heading, TranslatableText } from 'src/components/text';
+import useDashboardContext from './hooks/useDashboardContext';
+import styles from './DashboardPageV2.module.scss';
```
The PR for FE-1141 actually eliminates this styling, so no worries - we won't be incorporating new CSS modules here.

nicklemmon commented on src/pages/dashboard/DashboardPageV2.js:
```diff
@@ -0,0 +1,95 @@
+      </ScreenReaderOnly>
+
+      <Stack>
+        {currentUser?.first_name && (
```
🕺

nicklemmon commented on src/pages/dashboard/index.js:
```diff
@@ -1,3 +1,4 @@
 import DashboardPage from './DashboardPage.container';
```
Question for the reviewer - should I just move these files in to a new folder instead of co-locating here? `src/pages/dashboardV2`?

bkemper replied:
I would vote for making a new folder to avoid shared files and a much easier removal of the old dashboard when ready.

bkemper commented on src/pages/dashboard/DashboardPageV2.container.js:
```diff
@@ -0,0 +1,29 @@
+)(props => {
+  const [{ isHibanaEnabled }] = useHibana();
+
+  if (!isHibanaEnabled) return <Redirect to="/404" />;
```
`/404`?  Shouldn't we just handle this logic in the route condition?

bkemper commented on src/pages/dashboard/DashboardPageV2.container.js:
```diff
@@ -0,0 +1,29 @@
+  if (!isHibanaEnabled) return <Redirect to="/404" />;
+
+  return (
+    <DashboardContextProvider value={props}>
```
👏 Page State!

bkemper commented on src/pages/dashboard/DashboardPageV2.js:
```diff
@@ -0,0 +1,95 @@
+            <div className={styles.PanelRow}>
+              <Panel>
+                <Panel.Section>
+                  <PanelTitle>
```
Should we just add this to our shared matchbox component?

bkemper commented on src/pages/dashboard/DashboardPageV2.js:
```diff
@@ -0,0 +1,95 @@
+              <Panel>
+                <Panel.Section>
+                  <PanelTitle>
+                    <PanelIcon as={Code} />
```
Same here.

nicklemmon commented on src/components/matchbox/Banner.stories.mdx:
```diff
@@ -1,34 +1,23 @@
 import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
```
I was looking to document the new Panel components, but they don't work at the moment as Hibana is not enabled for storybook. We can work on getting that going at a later time though.

While I was there, I noticed some outdated documentation for the `<Banner/>` component so wanted to go ahead and update before I forgot.

bkemper replied:
Can we enable Hibana by default for the docs?

nicklemmon commented on src/components/matchbox/tests/Panel.test.js:
```diff
@@ -4,6 +4,7 @@ import { shallow } from 'enzyme';
 import { useHibana } from 'src/context/HibanaContext';
 jest.mock('src/context/HibanaContext');
 
+/* eslint-disable jest/expect-expect */
```
There were warnings littered throughout due to a series of re-usable functions that were written to test things.

nicklemmon commented on src/reducers/currentUser.js:
```diff
@@ -17,7 +17,18 @@ export default (state = initialState, { type, payload, meta }) => {
+        // TODO: Remove when OG theme is removed
+        // Just defaults the app state to having Hibana enabled when using the env variable
+        // Run the app locally with `npm run start-app-hibana`
+        ...(process.env.REACT_APP_DEFAULT_TO_HIBANA
```
Rather than manually overriding the context value, I refactored this to override the value in the Redux store. Needed to do it this way so that the condition helpers would work as on the route (`isUserUiOptionSet('isHibanaEnabled')`)

bkemper replied:
This is clever.  Wouldn't this work better as `initialState`?

nicklemmon replied:
I think the issue there is we use this flag primarily for Cypress testing - and unless we want to have a fixture that somehow dynamically overrides the value when Hibana is enabled, then when the user is successfully retrieved it would override initial state.

But let me try it! I suppose if the option isn't present in the fixture, then it won't override anything.

nicklemmon commented on cypress/tests/integration/accounts/billing/billingPage.spec.js:
```diff
@@ -500,7 +500,7 @@ describe('Billing Page', () => {
       it('closes the modal when clicking "Cancel"', () => {
         cy.findByLabelText('First Name').should('be.visible');
-        cy.findByRole('button', { name: 'Cancel' }).click();
+        cy.findByRole('button', { name: 'Cancel' }).click({ force: true });
```
Seeing some flakiness when clicking these buttons, so trying to use the `force` option to make sure it happens.

nicklemmon commented:
@bkemper - incorporated your feedback. All good stuff!

- I moved everything to a new folder
- Removed the `<Redirect />` and added the condition to the route - and as a result, I had to re-work `npm run start-app-hibana`
- Though UX indicated the larger panel headings and icons were dashboard-specific, we all know that might not last forever, so I made new `<Panel.Headline />` and `<Panel.HeadlineIcon />` global components.

bkemper commented on src/components/matchbox/Panel.js:
```diff
@@ -41,9 +44,47 @@ const Footer = props => {
+};
+
+HeadlineIcon.propTypes = {
+  as: PropTypes.func,
```
Can we use `React.element`?

nicklemmon replied:
For some reason that didn't work?

Do you mean:
```
HeadlineIcon.propTypes = {
  as: React.element,
}
```

That doesn't seem to do anything? Similarly `as: PropTypes.element` doesn't work - proptypes seems to register the icon component as a function.

bkemper commented on src/reducers/currentUser.js:
```diff
@@ -17,7 +17,18 @@ export default (state = initialState, { type, payload, meta }) => {
+        // Just defaults the app state to having Hibana enabled when using the env variable
+        // Run the app locally with `npm run start-app-hibana`
+        ...(process.env.REACT_APP_DEFAULT_TO_HIBANA
+          ? { options: { ...state.options, ui: { ...state.options.ui, isHibanaEnabled: true } } }
```
Hmmm... won't this override options being provided from `payload`?

nicklemmon replied:
Possibly - let me try the initial state thing and see how that goes.

bkemper commented on src/reducers/currentUser.js:
```diff
@@ -17,7 +17,26 @@ export default (state = initialState, { type, payload, meta }) => {
+          ...payload.options,
+          ui: {
+            ...state.options.ui,
+            ...payload.options.ui,
```
I believe there is one case when `options` is not provided — subaccount reporting user.
